core/trino-main/src/main/java/io/trino/execution/SqlTask.java

```diff
@@ -290,7 +292,16 @@ public Optional<Set<CatalogHandle>> getCatalogs()
     public boolean setCatalogs(Set<CatalogHandle> catalogs)
     {
         requireNonNull(catalogs, "catalogs is null");
-        return this.catalogs.compareAndSet(null, requireNonNull(catalogs, "catalogs is null"));
+        return this.catalogs.compareAndSet(null, catalogs);
+    }
+
+    public void setCatalogsLoaded(ListenableFuture<Void> catalogsLoadedFuture)
+    {
+        requireNonNull(catalogsLoadedFuture, "catalogsLoaded is null");
+        if (this.catalogsLoadedFuture.compareAndSet(null, catalogsLoadedFuture)) {
```

Comment:
throw an error if `compareAndSet` failed. Can it be called multiple times?

Reply:
yes. Update task can be called multiple times. Believe split assignments also come through this endpoint